commitizen/providers.py

```diff
@@ -0,0 +1,234 @@
+from __future__ import annotations
+
+import json
+import re
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, Callable, ClassVar, Optional, cast
+
+import importlib_metadata as metadata
+import tomlkit
+from packaging.version import VERSION_PATTERN, Version
+
+from commitizen.config.base_config import BaseConfig
+from commitizen.exceptions import VersionProviderUnknown
+from commitizen.git import get_tags
+
+ENTRYPOINT = "commitizen.provider"
+DEFAULT = "commitizen"
+
+
+class VersionProvider(ABC):
```

Comment:
What do you think of using protocols?

Reply:
In this precise case, I prefer `ABC` over `Protocol` because:
- I expect both a constructor and methods (I could move the `config` in the method signatures, but this makes more verbose code without benefits in my opinion)
- it is intended for user/community to extend this, using `ABC` force to implement methods and raise an explicit error if not, which I think is easier to understand (so more discoverable for them, less support for you)

But if you prefer `Protocol` over `ABC` I can switch 👍🏼

Reply:
@woile What do you think?

Reply:
I prefer `Protocol` 🤣 but using `ABC` is fine as well. Point 2 is the key, with `Protocols` people would have to use `mypy` to catch the problems, I like the explicit error.